torch/testing/_internal/common_methods_invocations.py

```diff
@@ -2794,21 +2805,25 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
                                 device_type='cuda', dtypes=[torch.cdouble], active_if=IS_WINDOWS),
                    )),
     OpInfo('addmm',
-           dtypes=floating_types(),
+           dtypes=floating_and_complex_types_and(torch.float16),
            dtypesIfCPU=all_types_and_complex_and(torch.float16, torch.bfloat16),
-           # BFloat16 support on CUDA requires CUDA 11 and SM53
-           dtypesIfCUDA=floating_types_and(torch.float16, torch.complex64, torch.complex128,
-                                           *[torch.bfloat16] if CUDA11OrLater else []),
-           dtypesIfROCM=floating_types_and(torch.float16, torch.complex64, torch.complex128, torch.bfloat16),
+           dtypesIfROCM=floating_and_complex_types_and(torch.float16, torch.bfloat16),
+           dtypesIfCUDA=floating_and_complex_types_and(torch.float16, *[torch.bfloat16] if CUDA11OrLater else []),
            assert_autodiffed=True,
-           autodiff_nonfusible_nodes=['aten::add', 'aten::mm'],
-           skips=(
-               # Skips unsupported bfloat16 check because above support check
-               #   doesn't work on all platforms
-               SkipInfo('TestOpInfo', 'test_unsupported_dtypes', dtypes=(torch.bfloat16,)),
-               # TODO: remove redundant method_tests() entries
-               SkipInfo('TestOpInfo', 'test_duplicate_method_tests')),
+           supports_inplace_autograd=False,
            sample_inputs_func=sample_inputs_addmm),
+    OpInfo('addmm',
+           # Another OpInfo is required for addmm because inputs whose alpha & beta are not 1
```

Comment:
I'm confused by this.

When alpha=beta=1 I would expect addmm to be replaced with a call to linear:

https://github.com/pytorch/pytorch/blob/c371542efc31b1abfe6f388042aa3ab0cef935f2/torch/csrc/jit/passes/fuse_linear.cpp#L36

But for this OpInfo the alpha=beta=1 but aten:add and aten::mm are described as "nonfusible"?

Let's dig in! cc @heitorschueroff @Lilyjjo @jerryzh168 

```
def foo(a, b, c):
  return torch.addmm(a, b, c, alpha=1, beta=1)
scripted_foo = torch.jit.script(foo)
scripted_foo.graph
: graph(%a.1 : Tensor,
      %b.1 : Tensor,
      %c.1 : Tensor):
  %6 : int = prim::Constant[value=1]()
  %7 : Tensor = aten::addmm(%a.1, %b.1, %c.1, %6, %6)
  return (%7)
```

So here's the graph for a function that always calls addmm with alpha=beta=1. But the graph isn't instantiated for any inputs, so let's make that happen:

```
scripted_foo.graph_for(torch.randn(2, 2), torch.randn(2, 2), torch.randn(2, 2))
: graph(%a.1 : Tensor,
      %b.1 : Tensor,
      %c.1 : Tensor):
  %3 : int = prim::Constant[value=1]()
  %6 : Tensor = prim::profile[profiled_type=Float(2, 2, strides=[2, 1], requires_grad=0, device=cpu)](%b.1)
  %7 : Tensor = prim::profile[profiled_type=Float(2, 2, strides=[2, 1], requires_grad=0, device=cpu)](%c.1)
  %4 : Tensor = aten::mm(%6, %7) # <string>:3:24
  %8 : Tensor = prim::profile[profiled_type=Float(2, 2, strides=[2, 1], requires_grad=0, device=cpu)](%a.1)
  %9 : Tensor = prim::profile[profiled_type=Float(2, 2, strides=[2, 1], requires_grad=0, device=cpu)](%4)
  %5 : Tensor = aten::add(%8, %9, %3) # <string>:3:17
  %10 : Tensor = prim::profile[profiled_type=Float(2, 2, strides=[2, 1], requires_grad=0, device=cpu)](%5)
   = prim::profile()
  return (%10)
```

So here we see that when the graph is instantiated and alpha=beta=1 (and alpha and beta are compile-time constants, more on this later) then the addmm call is decomposed to an mm followed by add. 

This appears to be done here:

https://github.com/pytorch/pytorch/blob/9b9e19a80895da321912cab1b69fd4c878fabf9a/torch/csrc/jit/passes/decompose_ops.cpp#L94

I guess the pass that converts addmm to linear is only applied in certain circumstances?

And that's how we get add and mm, and it's true that we don't have the technology to fuse add and mm together, so these are nonfusible nodes. This suggests, by the way, that it'd more accurate to describe this OpInfo as something like "addmm_decomposed" vs. "nonfusible." Because the addmm was originally a fusion of mm + add.

Now, for fun, let's see what happens when alpha and beta aren't compile-time constants:

```
def bar(a, b, c, alpha: float, beta: float):
  return torch.addmm(a, b, c, alpha=alpha, beta=beta)
scripted_bar = torch.jit.script(bar)
scripted_bar.graph_for(torch.randn(2, 2), torch.randn(2, 2), torch.randn(2, 2), 1, 1)
: graph(%a.1 : Tensor,
      %b.1 : Tensor,
      %c.1 : Tensor,
      %alpha.1 : float,
      %beta.1 : float):
  %6 : Tensor = prim::profile[profiled_type=Float(2, 2, strides=[2, 1], requires_grad=0, device=cpu)](%a.1)
  %7 : Tensor = prim::profile[profiled_type=Float(2, 2, strides=[2, 1], requires_grad=0, device=cpu)](%b.1)
  %8 : Tensor = prim::profile[profiled_type=Float(2, 2, strides=[2, 1], requires_grad=0, device=cpu)](%c.1)
  %5 : Tensor = aten::addmm(%6, %7, %8, %beta.1, %alpha.1) # <ipython-input-19-7ce16fcd7831>:2:9
  %9 : Tensor = prim::profile[profiled_type=Float(2, 2, strides=[2, 1], requires_grad=0, device=cpu)](%5)
   = prim::profile()
  return (%9)
```

And here we see that passing alpha and beta as non-constant arguments to the addmm call is sufficient for the JIT to refuse to decompose addmm.

So I think these OpInfos are correct, but the comment and name need a tweak. The comment for this OpInfo suggests that alpha and beta aren't 1, but actually the alpha=beta=1 for this OpInfo. So the comment would be something like, "When alpha=beta=1 as compile-time constants the JIT will decompose addmm into mm and add," and the variant_test_name updated to something like "decomposed" and the sample inputs updated to `partial(sample_inputs_addmm(alpha=1, beta=1)`.

Reply:
Thanks for the info, @mruberry!

I'll revise the code & comments based on your suggestions. I think I had included the comment with the wrong `addmm` OpInfo, but as you suggested, I'll include correct comments with both.

I'm not cognizant of JIT & am not sure what fusion means here but I had assumed that if `add` & `mm` are said to be _fused_, for example, they happen together (so that the `mm`'s output for each cell doesn't have to be loaded into the registers (up the cache hierarchy) once again in order to perform the addition step). So, now that you explained what's happening here, I'm confused as to why `mm` and `add` can't be fused for `alpha=1` & `beta=1`. 

It seems my assumption regarding the meaning of fusion was incorrect. I tried reading the [JIT overview](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/OVERVIEW.md), but I don't have knowledge of compilers yet, so I couldn't understand what fusion means.

Reply:
@Lilyjjo @jerryzh168, 
Please confirm if the JIT overview the best way to get started on PyTorch JIT's code-generation part? Thanks!

Reply:
Sorry, by "fusion" in the above I just meant that the two operations appear as one operation. The actual implementation of the operation may or may do anything special.

Reply:
Thanks, I'll read up on Torchscript & JIT.

BTW, I'll try porting `addmm` to structured tonight.

Reply:
This comment thread has been fun to read! The JIT overview is the best documentation we have for getting started on JIT's code generation. 

This blog has deep dive articles on various parts of the JIT internals that are helpful to get some more information on what the JIT code is doing (the JIT related blog posts have JIT in the title): http://lernapparat.de/. 

For getting started on compilers, 'Engineering a Compiler' by Keith Cooper Linda Torczon is great, but it might be overkill if you're just wanting to contribute to JIT specifically.

Reply:
@Lilyjjo: Thanks for the info! That was very helpful.